[FEATURE]  Add PersistentVolumeClaimRetentionPolicy to GeneralConfig

vanchonlee opened the pull request:
### Description
The `PersistentVolumeClaimRetentionPolicy` feature allows you to control what happens to Persistent Volume Claims (PVCs) when a StatefulSet is deleted or when a pod in a StatefulSet is removed. This is particularly useful for OpenSearch clusters where you might want to control data retention when scaling down or deleting clusters.



### Issues Resolved
https://github.com/opensearch-project/opensearch-k8s-operator/issues/962

### Check List
- [x] Commits are signed per the DCO using --signoff 
- [ ] Unittest added for the new/changed functionality and all unit tests are successful
- [ ] Customer-visible features documented
- [x] No linter warnings (`make lint`)

If CRDs are changed:
- [x] CRD YAMLs updated (`make manifests`) and also copied into the helm chart
- [x] Changes to CRDs documented

Please refer to the [PR guidelines](https://github.com/opensearch-project/opensearch-k8s-operator/blob/main/docs/developing.md#submitting-a-pr) before submitting this pull request.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rootxrishabh reviewed (approved):
LGTM! Thanks for the PR

prudhvigodithi commented on opensearch-operator/pkg/builders/cluster.go:
```diff
@@ -457,7 +462,8 @@ func NewSTSForNodePool(
 			},
-			PodManagementPolicy: appsv1.OrderedReadyPodManagement,
+			PodManagementPolicy:                  appsv1.OrderedReadyPodManagement,
+			PersistentVolumeClaimRetentionPolicy: persistentVolumeClaimRetentionPolicy,
```
Thanks, I'm curious what would be the default when `persistentVolumeClaimRetentionPolicy` is not passed? Can we please handle that as well.

prudhvigodithi commented:
Thanks @vanchonlee for your contribution, this is a good change we can target this Release v2.8.0 https://github.com/opensearch-project/opensearch-k8s-operator/issues/1011.
@getsaurabh02 @rootxrishabh

vanchonlee commented:
> Thanks @vanchonlee for your contribution, this is a good change we can target this Release v2.8.0 #1011. @getsaurabh02 @rootxrishabh

I noticed this PR wasn’t included in the v2.8.0 release. Is there anything I can do to help resolve any issues?

josedev-union commented:
@vanchonlee could you resolve the conflicts?

vanchonlee commented:
> @vanchonlee ping on this - we are gearing up for a release and if you want this to be included lets fix conflicts and get it reviewed

I have fixed,

josedev-union commented:
> > @vanchonlee ping on this - we are gearing up for a release and if you want this to be included lets fix conflicts and get it reviewed
> 
> I have fixed,

@vanchonlee let's make this landed in 3.x release
Could you fix DCO first?
And make CRD changes in opensearch.org instead of the deprecated opster.io? https://github.com/opensearch-project/opensearch-k8s-operator/tree/main/opensearch-operator/api/opensearch.org/v1

vanchonlee commented:
> > > @vanchonlee ping on this - we are gearing up for a release and if you want this to be included lets fix conflicts and get it reviewed
> > 
> > 
> > I have fixed,
> 
> @vanchonlee let's make this landed in 3.x release Could you fix DCO first? And make CRD changes in opensearch.org instead of the deprecated opster.io? https://github.com/opensearch-project/opensearch-k8s-operator/tree/main/opensearch-operator/api/opensearch.org/v1

Sure, I'm on it. I'll update and complete the PR shortly.

vanchonlee committed:
feat: add PersistentVolumeClaimRetentionPolicy to GeneralConfig

Fix DCO sign-off and move CRD changes from deprecated opster.io API to
opensearch.org/v1 for the 3.x release.

Signed-off-by: vanchonlee <vanchonlee@gmail.com>

vanchonlee committed:
docs: document claim retention policy in CRD design

Signed-off-by: vanchonlee <vanchonlee@gmail.com>
